src/Agent.Worker/DiagnosticLogManager.cs

```diff
@@ -0,0 +1,262 @@
+            {
+                // The format of the logs is:
+                // Worker_20171003-143110-utc.log
+                DateTime fileCreateTime = DateTime.ParseExact(s: file.Name.Substring(startIndex: 7, length: 15), format: "yyyyMMdd-HHmmss", provider: CultureInfo.InvariantCulture);
+
+                if (fileCreateTime >= jobStartTimeUtc)
+                {
+                    workerLogFiles.Add(file.FullName);
+                }
+            }
+            
+            return workerLogFiles;
+        }
+
+        private string GetEnvironmentContent(int agentId, string agentName, ReadOnlyCollection<TaskInstance> tasks)
+        {
+            var builder = new StringBuilder();
+
+            builder.AppendLine($"Environment file created at(UTC): {DateTime.UtcNow}"); // TODO: Format this like we do in other places.
+            builder.AppendLine($"Agent Version: {Constants.Agent.Version}");
+            builder.AppendLine($"Agent Id: {agentId}");
+            builder.AppendLine($"Agent Name: {agentName}");
+            builder.AppendLine($"OS: {System.Runtime.InteropServices.RuntimeInformation.OSDescription}");
+            builder.AppendLine("Tasks:");
+
+            foreach (TaskInstance task in tasks)
+            {
+                builder.AppendLine($"\tName: {task.Name} Version: {task.Version}");
+            }
+
+#if OS_WINDOWS
+            // windows defender on/off
+            builder.AppendLine($"Defender enabled: {IsDefenderEnabled()}");
+
+            // firewall on/off
+            builder.AppendLine($"Firewall enabled: {IsFirewallEnabled()}");
+#endif
+
+            return builder.ToString();
+        }
+
+#if OS_WINDOWS
+        // Returns whether or not Windows Defender is running.
+        private static bool IsDefenderEnabled()
+        {
+            return Process.GetProcessesByName("MsMpEng.exe").FirstOrDefault() != null;
+        }
+#endif
+
+#if OS_WINDOWS
+        // Returns whether or not the Windows firewall is enabled.
+        private static bool IsFirewallEnabled()
+        {
+            try 
+            {
+                using (RegistryKey key = Registry.LocalMachine.OpenSubKey("System\\CurrentControlSet\\Services\\SharedAccess\\Parameters\\FirewallPolicy\\StandardProfile")) 
+                {
+                    if (key == null) { return false; } 
+
+                    Object o = key.GetValue("EnableFirewall");
+                    if (o == null) { return false; } 
+
+                    int firewall = (int) o;
+                    if (firewall == 1) { return true; } 
+                    return false;
+                }
+            } 
+            catch 
+            {
+                return false;
+            }
+        }
+#endif
+
+        private class DiagnosticLogMetadata
+        {
+            public DiagnosticLogMetadata(string agentName, string agentId, int poolId, string phaseName, string fileName, string phaseResult)
```

Comment:
you can put this into webapi, it will generate the typescript client as well.

Reply:
In progress.